core/blockchain.go

```diff
@@ -616,6 +616,136 @@ func (bc *BlockChain) SetSafe(header *types.Header) {
+				return bc.genesisBlock.Header(), rootNumber
+			}
+			headBlock = parent
+
+			// If the genesis block is reached, stop searching.
+			if headBlock.Number.Uint64() == 0 {
+				log.Info("Genesis block is reached", "number", headBlock.Number, "hash", headBlock.Hash())
+				return headBlock, rootNumber
+			}
+			continue // keep rewinding
+		}
+		// Once the available state is found, ensure that the requested root has already
+		// been crossed. If not, continue rewinding.
+		if beyondRoot || headBlock.Number.Uint64() == 0 {
+			log.Debug("Rewound to block with state", "number", headBlock.Number, "hash", headBlock.Hash())
+			return headBlock, rootNumber
+		}
+		log.Debug("Skipping block with threshold state", "number", headBlock.Number, "hash", headBlock.Hash(), "root", headBlock.Root)
+		headBlock = bc.GetHeader(headBlock.ParentHash, headBlock.Number.Uint64()-1) // Keep rewinding
+	}
+}
+
+// rewindPathHead implements the logic of rewindHead in the context of path scheme.
+func (bc *BlockChain) rewindPathHead(root common.Hash) (*types.Header, uint64) {
+	var (
+		beyondRoot = root == common.Hash{}            // Flag whether we're beyond the requested root (no root, always true)
+		headBlock  = bc.CurrentBlock()                // Head block of the chain
+		pivot      = rawdb.ReadLastPivotNumber(bc.db) // Associated block number of pivot point state
+		rootNumber uint64                             // Associated block number of requested root
+	)
+	// Rewind the head block tag until an available state is found.
+	for !bc.HasState(headBlock.Root) {
+		// If a root threshold was requested but not yet crossed, check
+		if !beyondRoot && headBlock.Root == root {
+			beyondRoot, rootNumber = true, headBlock.Number.Uint64()
+		}
+		// If pivot block is reached, return the genesis block as the
+		// new chain head. Theoretically there must be a persistent
+		// state before or at the pivot block, prevent endless rewinding
+		// towards the genesis just in case.
+		if pivot != nil && *pivot >= headBlock.Number.Uint64() {
+			log.Info("Pivot block is reached, aiming genesis", "number", headBlock.Number, "hash", headBlock.Hash())
+			return bc.genesisBlock.Header(), rootNumber
+		}
+		headBlock = bc.GetHeader(headBlock.ParentHash, headBlock.Number.Uint64()-1) // Keep rewinding
+	}
+	// The requested root has already been crossed. Return the block
+	// corresponding to the available state as the new chain head.
+	if beyondRoot {
+		return headBlock, rootNumber
+	}
+	// If the requested root has not yet been crossed, prevent further
+	// backward rewinding if reaching the root is deemed impossible.
+	if !bc.HasState(root) && !bc.stateRecoverable(root) {
+		return headBlock, 0
+	}
+	// Rewind the head block tag until the requested root is reached.
+	for !beyondRoot {
+		headBlock = bc.GetHeader(headBlock.ParentHash, headBlock.Number.Uint64()-1) // Keep rewinding
+		beyondRoot = headBlock.Root == root
+	}
```

Comment:
What if the loop at line `697` ended so that  `root == headBlock.Root`, `bc.HasState(root) == true` ? 
If it exits the loop due to `bc.HasState(root)`, then we do not update the variable `beyondRoot`, so it's still `false`. 

If that hppens, this clause will rewind it one step further, and it looks to me like it will then make things break pretty badly: if we miss the root with state, we're very much of luck, no?

Reply:
Ah, right, thanks for catching up